nanopass/nanopass.nim

```diff
@@ -0,0 +1,445 @@
+      else:
+        error("expected '|'", n[0])
+    of nnkCall, nnkIdent:
+      list.add n
+    else:
+      error("unexpected syntax: " & $n.kind, n)
+
+  for it in body.items:
+    case it.kind
+    of nnkInfix:
+      if it[0].eqIdent("::="):
+        var nt = NonTerminalDef(name: it[1])
+        extract(it[2], nt.add)
+        def.add nt
+        continue
+    of nnkCall:
+      add.add it
+      continue
+    else:
+      discard "report an error below"
+
+    error("items must be of the form `a in b`, or `a ::= ...`", it)
+
+  # to keep the implementation simple, a non-extension language is treated
+  # internally as an empty language definition being extended
+  buildLanguage(add, @[], def, default(LangDef), body)
+
+proc makeLanguage(base: LangDef, body: NimNode): LangDef =
+  ## Creates a language definition from the ``defineLanguage`` DSL code
+  ## and `base`.
+  var add, sub: seq[NimNode]
+  var def: seq[NonTerminalDef]
+
+  proc extract(n: NimNode, add, sub: var seq[NimNode]) =
+    case n.kind
+    of nnkPrefix:
+      if n[0].eqIdent("+"):
+        add.add n[1]
+      elif n[0].eqIdent("-"):
+        sub.add n[1]
+      else:
+        error("expected `+` or `-`", n[0])
+    of nnkInfix:
+      if n[0].eqIdent("|"):
+        n.expectLen 3
+        extract(n[1], add, sub)
+        extract(n[2], add, sub)
+      else:
+        error("expected '|'", n[0])
+    else:
+      error("unexpected syntax: " & $n.kind, n)
+
+  for it in body.items:
+    var handled = false
+    case it.kind
+    of nnkInfix:
+      if it[0].eqIdent("::="):
+        it.expectLen 3
+        var nt = NonTerminalDef(name: it[1])
+        nt.name.expectKind nnkCall
+        extract(it[2], nt.add, nt.sub)
+        def.add nt
+        handled = true
+    of nnkPrefix:
+      if it[0].eqIdent("-"):
+        sub.add it[1]
+        handled = true
+      elif it[0].eqIdent("+"):
+        add.add it[1]
+        handled = true
+    of nnkCall:
+      # non-terminal with no change in productions
+      def.add NonTerminalDef(name: it)
+      handled = true
+    else:
+      discard
+
+    if not handled:
+      error("expected `-a`, `+a`, or `a(...) ::= ...`", it[0])
+
+  buildLanguage(add, sub, def, base, body)
+
+macro defineLanguage*(name, body: untyped) =
+  ## Creates a language definitions and binds it to a const symbol with the
+  ## given name.
+  body.expectKind nnkStmtList
+  body.expectMinLen 1
+  let p = bindSym"makeLanguage"
+  let q = bindSym"quote"
+  if body[0].kind == nnkCommentStmt:
+    let filtered = body[1..^1]
+    result = nnkConstSection.newTree(
+      nnkConstDef.newTree(name,
+        newEmptyNode(),
+        quote do: `p`(`q` do: `filtered`)),
+      body[0])
+  else:
+    result = nnkConstSection.newTree(
+      nnkConstDef.newTree(name,
+        newEmptyNode(),
+        quote do: `p`(`q` do: `body`)))
+
+macro defineLanguage*(name, base, body: untyped) =
+  ## Creates a language definition, extending `base`, and binds it to a const
```

Comment:
Is the intention, that after nanopass is implemented, to look at what it would mean to go with an extension/inheritance vs composition based API?

Reply:
Yep, I want to have something that's certain to work first, which is also the reason metadata is not part of the PR.